p2p/discovery/bucket.go

```diff
@@ -30,24 +36,67 @@ func (b *bucket) addNode(node *Node) {
 		// do nothing for now
 		// TODO lru
 	} else {
+		b.lock.Lock()
+		defer b.lock.Unlock()
+
+		log.Error("add node: %s", hexutil.BytesToHex(node.ID.Bytes()))
 		if len(b.peers) < bucketSize {
 			b.peers = append(b.peers, node)
 		} else {
 			copy(b.peers[:], b.peers[1:])
 			b.peers[len(b.peers)-1] = node
 		}
 	}
-
-	log.Debug("bucket size: %d", len(b.peers))
 }
 
 // hasNode check if the bucket already have this node, if so, return its index, otherwise, return -1
 func (b *bucket) hasNode(node *Node) int {
+	b.lock.Lock()
+	defer b.lock.Unlock()
 	for index, n := range b.peers {
-		if n.sha == node.sha {
+		if n.ID == node.ID {
 			return index
 		}
 	}
 
 	return -1
 }
+
+func (b *bucket) deleteNode(target *common.Hash) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	index := -1
+	for i, n := range b.peers {
+		sha := n.ID.ToSha()
+		if *sha == *target {
+			index = i
+			break
+		}
+	}
+
+	if index == -1 {
+		log.Error("don't find the node to delete\n")
```

Comment:
"Failed to find the node to be deleted\n" would be better.

Reply:
good suggestion